extra_files/fv3jedi_ufs_mod.F90

```diff
@@ -0,0 +1,585 @@
+! (C) Copyright 2020 NOAA
+!
+! This software is licensed under the terms of the Apache Licence Version 2.0
+! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+
+#define esmf_err_abort(rc) if (esmf_LogFoundError(rc, msg="Aborting UFS", line=__LINE__, file=__FILE__)) call esmf_Finalize(endflag=esmf_END_ABORT)
+
+module fv3jedi_ufs_mod
+  
+  ! oops
+  use datetime_mod
+  use duration_mod
+
+  ! fckit
+  use fckit_configuration_module, only: fckit_configuration
+  
+  ! fv3jedi
+  use fv3jedi_geom_mod,      only: fv3jedi_geom
+  use fv3jedi_state_mod,     only: fv3jedi_state
+  
+  ! ufs
+  use ESMF
+  use NUOPC
+  use NUOPC_Driver
+  use module_EARTH_GRID_COMP, only: esmSS => EARTH_REGISTER
+  use mpp_mod,            only: read_input_nml,mpp_pe
+  
+
+  implicit none
+  private
+  
+  public :: model_ufs
+  
+  !> Fortran derived type to hold model definition
+  type :: model_ufs
+     type(ESMF_GridComp) :: esmComp
+     type(ESMF_State) :: toJedi, fromJedi 
+     type(esmf_Clock) :: clock
+     type(esmf_config) :: cf_main                                         !<-- the configure object
+   contains
+     procedure :: create
+     procedure :: delete
+     procedure :: initialize
+     procedure :: step
+     procedure :: finalize
+  end type model_ufs
+  
+  character(len=*), parameter :: modname='fv3jedi_ufs_mod'
+  
+  ! --------------------------------------------------------------------------------------------------
+
+contains
+  
+  ! --------------------------------------------------------------------------------------------------
+  
+  subroutine create(self, conf, geom)
+    
+    implicit none
+    class(model_ufs),          intent(inout) :: self
+    type(fckit_configuration), intent(in)    :: conf
+    type(fv3jedi_geom),        intent(in)    :: geom
+    
+    integer :: rc, urc, phase, i, cnt
+    character(len=20) :: cdate_start, cdate_stop
+
+    type(ESMF_Time)         :: startTime, stopTime
+    type(ESMF_TimeInterval) :: timeStep
+
+    character(len=*),parameter :: subname = modname//' (create)'
+    type(ESMF_CplComp),  pointer       :: connectors(:)
+    character(len=128) :: name, msg
+    
+    ! Initialize ESMF
+    call ESMF_Initialize(logkindflag=esmf_LOGKIND_MULTI, &
+         defaultCalkind=esmf_CALKIND_GREGORIAN, &
+         mpiCommunicator=geom%f_comm%communicator(), rc=rc)
+    esmf_err_abort(rc)
+    ! Flush log output while debugging
+    call ESMF_LogSet(flush=.true., rc=rc)
+    esmf_err_abort(rc)
+    call ESMF_LogWrite("ESMF Initialized in "//subname, ESMF_LOGMSG_INFO)
+
+    self%cf_main=esmf_configcreate(rc=rc)   
+    call ESMF_ConfigLoadFile(config=self%cf_main, &
+         filename='model_configure', &
+         rc=rc)
+    
+    ! This call to read_input_nml() seems to be required
+    ! for CCPP.  However, it does not belong at this level
+    ! but should be handled inside the model itself
+    call read_input_nml()
```

Comment:
Need to figure out what to do with this part.  If `model_configure` is a config file for the ufs atmosphere, then it should be read in at that level.  If "earth-wide" configuration is needed (it probably is -- e.g., coupled restarts) then a separate configuration file is needed.  Is this `nems.configure?`